save-frontend/src/main/kotlin/com/saveourtool/save/frontend/utils/CustomHooks.kt

```diff
@@ -152,3 +161,153 @@ fun useBackground(colorStyle: Style) {
+ *   should be allowed, the default is `false`.
+ * @param eventType the event type selector.
+ *   The same HTTP endpoint may return events of different types, and this
+ *   selector allows to receive only a certain subset of the whole event volume.
+ *   The default is `message`.
+ *   On the server side, use `org.springframework.http.codec.ServerSentEvent` to
+ *   create an event with a custom type.
+ * @param init invoked before an event stream is requested.
+ *   Allowed to change the state of components.
+ * @param onCompletion invoked if the response is `HTTP 200 OK`, and when the
+ *   server closes the connection.
+ *   Allowed to change the state of components.
+ * @param onError invoked when the event source reports an error.
+ *   Allowed to change the state of components.
+ * @param onEvent invoked when a new event arrives.
+ *   Allowed to change the state of components.
+ * @return a lambda that triggers the callback.
+ * @see useNewlineDelimitedJson
+ */
+@Suppress(
+    "LongParameterList",
+    "TOO_MANY_PARAMETERS",
+    "TYPE_ALIAS",
+)
+fun useEventStream(
+    url: String,
+    withCredentials: Boolean = false,
+    eventType: String = "message",
+    init: EffectBuilder.() -> Unit = {},
+    onCompletion: EffectBuilder.() -> Unit = {},
+    onError: EffectBuilder.(error: Event, readyState: Short) -> Unit = { _, _ -> },
+    onEvent: EffectBuilder.(message: MessageEvent) -> Unit,
+): () -> Unit =
+        useDeferredEffect {
+            init()
+
+            val source = EventSource(
+                url = url,
+                eventSourceInitDict = jso {
+                    this.withCredentials = withCredentials
+                },
+            )
+
+            source.addEventListener(eventType, { event ->
+                onEvent(event as MessageEvent)
+            })
+
+            source.onerror = { error ->
+                if (source.readyState == CONNECTING) {
+                    source.close()
+                    onCompletion()
+                } else {
+                    onError(error, source.readyState)
+                }
+            }
+        }
+
+/**
+ * Reads the response of `application/x-ndjson` `Content-Type`.
+ *
+ * Only works inside functional components.
+ *
+ * @param url the URL that accepts an `HTTP GET` and can respond with a
+ *   `application/x-ndjson` `Content-Type`.
+ * @param init invoked before an event stream is requested.
+ *   Allowed to change the state of components.
+ * @param onCompletion invoked if the response is `HTTP 200 OK`, and when the
+ *   server closes the connection.
+ *   Allowed to change the state of components.
+ * @param onError invoked when the event source reports an error.
+ *   Allowed to change the state of components.
+ * @param onEvent invoked when a new event arrives.
+ *   Allowed to change the state of components.
+ * @return a lambda that triggers the callback.
+ * @see useEventStream
+ */
+internal fun useNewlineDelimitedJson(
```

Comment:
maybe name it as `Jsonl`:
```suggestion
internal fun useJsonl(
```

Reply:
Addressed with e5f501d371cbcad8f71880d4780555fc8fb2e1d2